### DISCUSSION TITLE
Duplex channels over NetHttpBinding (websockets)

### DISCUSSION THREAD

**Discussion by absinghal** [CoreWCF/CoreWCF#1021] [Q&A]
Hello

Just want to confirm whether Duplex channels are supported over NetHttpBinding (websockets)

Can see code in repo that checks for web socket connections but when I tried in my code, getting error as follows on client

`WebSocketException: WebSocket upgrade request failed. Received response status code '200 (OK)', expected: '101 (SwitchingProtocols)'`

Also the NetCoreServer sample doesn't seem to have code that shows how to enable NetHttpBinding with duplex channels

Kindly advise

**Comment by absinghal** [comment]
@mconnew can you kindly comment on this. Thanks.

**Reply by samkeusch** [reply]
I got the same Issue. In my case it seams to be the problem that the WSDL address and websocket are the same (legacy from the .Net Framework 4.8 project I am trying to migrate). My workaround/solution for this at the moment is that I configure the websocket endpoint to some other address and for backwards compatibility I route websocket upgrade requests to the new address.

**Comment by absinghal** [comment]
Seems like I found the reason.

MetaDataMiddleware is causing the WebSocketException

Under MetadataServiceCollectionExtensions, MetadataTransportServiceBuilder is inserted purposefully before any ITransportServiceBuilder, which includes HttpTransportServiceBuilder.

Hence ServiceModelHttpMiddleware, where a branched pipeline issues AcceptWebSockets doesnt get a chance to execute first. Instead MetaDataMiddleware executes, and hence the exception.

Workaround for now: Dont use ServiceModelMetaData, if websockets/duplex nethttpbinding is to be used.

Can comment lines like following:
 //services.AddServiceModelMetadata().
     //AddSingleton<IServiceBehavior, UseRequestHeadersForMetadataAddressBehavior>();

**Comment by absinghal** [comment]
@samkeusch can you provide more details on how you implemented "My workaround/solution for this at the moment is that I configure the websocket endpoint to some other address and for backwards compatibility I route websocket upgrade requests to the new address"

**Reply by samkeusch** [reply]
Sure
My requirement is to be able to call for example https://localhost/myservice/websocket and https://localhost/myservice/websocket?wsdl
So what i did is to route the calls to https://localhost/myservice/websocket to a different address for example https://localhost/myservice/websocket/xyz (I use a Startup Class):
```
public void Configure(IApplicationBuilder app)
{
    app.Use(async (context, next) =>
    {
       // (WebSocket Upgrade Requests return "The server returned status code '200' when status code '101' was expected. " )
       if ( IsItTheAddressToRoute(context) )
       {
           if (context.Request.Headers.Contains(sHttpRequestWebSocketUpgradeHeader)) // new KeyValuePair<string, StringValues>("Upgrade", "websocket")
            context.Request.Path += "/xyz";
       }
    await next.Invoke();
   });

   app.UseServiceModel(serviceBuilder =>
   {
      ...
   }

}
```

**Reply by absinghal** [reply]
Thats a great approach.

I think we could add this to MetaDataMiddleware InvokeAsync method.

If the request header has the upgrade request, the MetaDataMiddleware should simply invoke the next middleware, which would likely be the  ServiceModelHttpMiddleware. 

@mconnew Kindly advise, thanks.

**Reply by mconnew** [reply]
The upgrade header isn't only used for WebSockets, it can be used for upgrading an HTTP/1.1 connection to HTTP/2. I don't know if that request would get this far or would it be handled earlier in the pipe? I don't like fixing bugs twice so want to make sure we have the right constraints here and do the right thing for all requests with the Upgrade header.

**Reply by mconnew** [reply]
I checked with a dev in asp.net core and we need to worry about HTTP/2 requests that upgrade to WebSockets as they look slightly different. Also other upgrade mechanisms, eg HTTP/1.1 to HTTP/2 aren't supported by at least Kestrel, and there's no API surface in asp.net core to facilitate that. So I think we can safely only worry about WebSockets requests, and there's a cleaner way to do that. Simply check `context.WebSockets.IsWebSocketRequest`. The fix would be to change `MetadataMiddleware.BuildBranch` to have the short circuit code which skips the middleware if the request isn't a `GET` request to look like this:
```c#
    if ("GET".Equals(reqContext.Request.Method, StringComparison.OrdinalIgnoreCase) && !reqContext.WebSockets.IsWebSocketRequest)
    {
        // If request is a GET request, continue on the branchApp middleware chain
        // to handle requests for WSDL, HelpPage etc
        return branchNext(reqContext);
    }
```

**Reply by absinghal** [reply]
Awesome.

Looks like a highly efficient and effective fix.

**Comment by absinghal** [comment]
Hello @mconnew do you need a PR for this ?

**Reply by mconnew** [reply]
If you would be willing to do that (along with a test which fails without the fix), it would be greatly appreciated and will result in a fix being released faster.

**Reply by absinghal** [reply]
I can try.

Can you give an idea what needs to be done for the test part ?

Thanks

**Reply by mconnew** [reply]
Add a test to the CoreWCF.Http.Tests project which sets up a WebSockets endpoint and also accepts GET requests at the same path. The reported issue was for fetching the WSDL at the same path. Then use HttpClient to fetch the WSDL and see that it fails. Then apply the fix, and the test should now pass. It should be easy enough looking at existing tests to see the pattern for creating a new test.

**Comment by BrianDT** [comment]
Did this fix go in? and when will it be released? Assuming its not in yet because I'm experiencing the same issue in version 1.8.0
